source/common/upstream/health_discovery_service.cc

```diff
@@ -252,13 +262,68 @@ void HdsCluster::startHealthchecks(AccessLog::AccessLogManager& access_log_manag
 
 void HdsCluster::initialize(std::function<void()> callback) {
   initialization_complete_callback_ = callback;
-  for (const auto& host : *initial_hosts_) {
+}
+
+void HdsCluster::update(const envoy::api::v2::Cluster& cluster) {
+  // Update endpoints
+  HostVector added_hosts;
+  HostVector removed_hosts;
+  HostVectorSharedPtr current_hosts(new HostVector(*hosts_)); // Not sure if this copy is necessary
+  HostVectorSharedPtr final_hosts(new HostVector());
+  HostMap updated_hosts;
+  std::unordered_set<std::string> existing_hosts(hosts_->size());
+
+  for (const auto& address : cluster.hosts()) {
+    const HostSharedPtr& host = HostSharedPtr{
+        new HostImpl(info_, "", Network::Address::resolveProtoAddress(address),
+                     envoy::api::v2::core::Metadata::default_instance(), 1,
+                     envoy::api::v2::core::Locality().default_instance(),
+                     envoy::api::v2::endpoint::Endpoint::HealthCheckConfig().default_instance(), 0,
+                     envoy::api::v2::core::HealthStatus::UNKNOWN)};
+
     host->healthFlagSet(Host::HealthFlag::FAILED_ACTIVE_HC);
+
+    if (updated_hosts.count(host->address()->asString())) {
+      continue;
+    }
+
+    auto existing_host = all_hosts_.find(host->address()->asString());
+    if (existing_host != all_hosts_.end()) {
+      // TODO: update endpoint health check port
+      existing_hosts.emplace(host->address()->asString());
+      updated_hosts[host->address()->asString()] = existing_host->second;
+      final_hosts->push_back(existing_host->second);
+    } else {
+      added_hosts.push_back(host);
+      updated_hosts[host->address()->asString()] = host;
+      final_hosts->push_back(host);
+    }
+  }
+
+  // Remove found hosts from current hosts to get removed hosts
+  for (auto itr = current_hosts->begin(); itr != current_hosts->end();) {
+    auto existing_itr = existing_hosts.find((*itr)->address()->asString());
+
+    if (existing_itr != existing_hosts.end()) {
+      existing_hosts.erase(existing_itr);
+      itr = current_hosts->erase(itr);
+    } else {
+      itr++;
+    }
+  }
```

Comment:
Is this modeled after similar code in `BaseDynamicClusterImpl::updateDynamicHostList`? If so, please comment.

Reply:
Yeah, do you mean I should add a comment in the code mentioning `BaseDynamicClusterImpl::updateDynamicHostList`?

Reply:
Yep.